src/interface/block_chain.cpp

```diff
@@ -293,6 +299,251 @@ void block_chain::push(transaction_const_ptr tx, dispatcher&,
+
+}
+
+// Once a transaction was mined, every previous output that was marked
+// as TEMPORARY SPENT, needs to be removed
+// (since it was permanetly added to the spent database)
+void block_chain::remove_spend(libbitcoin::hash_digest hash){
+    chosen_spent_.erase(hash);
+}
+
+// Search the TEMPORARY SPENT index for conflicts between previous output
+std::set<libbitcoin::hash_digest> block_chain::get_double_spend_chosen_list(transaction_const_ptr tx) {
+
+    std::set<libbitcoin::hash_digest> spent_conflict{};
+    for (auto const& input : tx->inputs()) {
+        for(auto const& spents : chosen_spent_){
+            for(auto const& previous : spents.second){
+                if(previous.output_hash == input.previous_output().hash() &&
+                    previous.output_index == input.previous_output().index())
+                {
+                    spent_conflict.insert(tx->hash());
+                }
+            }
+        }
+    }
+    return spent_conflict;
+}
+
+// Search the spent database and the TEMPORARY SPENT for conflicts
+bool block_chain::check_is_double_spend(transaction_const_ptr tx){
+    bool is_double_spend = false;
+    for(auto const& input : tx->inputs()){
+        boost::latch latch(2);
+        fetch_spend(input.previous_output(), [&](const libbitcoin::code &ec, libbitcoin::chain::input_point input) {
+            if (ec == libbitcoin::error::success) {
+                is_double_spend = true;
+            }
+            latch.count_down();
+        });
+        latch.count_down_and_wait();
+
+        if(is_double_spend) return true;
+    }
+    return !get_double_spend_chosen_list(tx).empty();
+}
+
+
+// Insert to chosen transaction List, ordered by the benefit yielded by the transaction
+bool block_chain::insert_to_chosen_list(transaction_const_ptr& tx, double benefit, size_t tx_size, size_t tx_sigops){
+#ifdef BITPRIM_CURRENCY_BCH
+    tx_benefit txnew{benefit, tx_sigops, tx_size, tx->fees(), tx->to_data(1), tx->hash()};
+#else
+    tx_benefit txnew{benefit, tx_sigops, tx_size, tx->fees(), tx->to_data(1), tx->hash(), tx->hash(true)};
+#endif
+    chosen_unconfirmed_.insert(
+        std::upper_bound(begin(chosen_unconfirmed_), end(chosen_unconfirmed_), txnew, [](tx_benefit const& a, tx_benefit const& b) {
+            return a.benefit < b.benefit;
+    }), txnew);
+
+    chosen_size_ += tx_size;
+    chosen_sigops_ += tx_sigops;
+    append_spend(tx);
+
+    return true;
+}
+
+//Search how many transactions we need to remove from the transaction chosen list, if we want to add a new transaction
+size_t block_chain::find_txs_to_remove_from_chosen(const size_t sigops_limit, const size_t tx_size,
+        const size_t tx_sigops, const size_t tx_fees, const double benefit,
+            size_t& acum_sigops, size_t& acum_size, double& acum_benefit)
+{
+    // How many transactions should I remove, while still gaining more benefit
+    acum_benefit += (*chosen_unconfirmed_.begin()).benefit;
```

Comment:
if the container is empty this will fail, are you sure the `chosen_unconfirmed_` is not empty?

Reply:
This method is only called when the chosen_unconfirmed_ list is full